modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/util/UCharReplacer.java

```diff
@@ -0,0 +1,206 @@
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+import com.ibm.icu.text.Normalizer2;
+import com.ibm.icu.text.UCharacterIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+/**
+ * Substitute unicode characters with replacement strings.
+ *
+ * <p>The substitution is made using different attempts in a series of preference;
+ * each successor is considered a fallback to its predecessor.</p>
+ * <ol>
+ *   <li>Locate a replacement string in one or several user provided tables;</li>
+ * </ol>
+ * 
+ * <p>The use of this class <em>may</em> result in a change in unicode character composition between input and output.
+ * If you need a certain normalization form, normalize after the use of this class.</p>
+ * 
+ * <p>Usage example:</p>
+ * <code><pre>
+ * 	UCharReplacer ucr = new UCharReplacer();
+ *	ucr.addTranslationTable(fileURL);
+ *  ucr.addTranslationTable(fileURL2);
+ *	String ret = ucr.toReplacementString(input);
+ * </pre></code>
+ *  
+ * <p>The translation table file is using the same xml format as that of
+ * java.util.Properties [1][2], using the HEX representation (without 
+ * the characteristic 0x-prefix!) of a unicode character as the <tt>key</tt> 
+ * attribute and the replacement string as value of the <tt>entry</tt> 
+ * element.</p>
+ * 
+ * <p>If the <tt>key</tt> attribute contains exactly one unicode codepoint 
+ * (one character) it will be treated literally. It will not be interpreted 
+ * as a HEX representation of another character, even if theoretically 
+ * possible. E.g. if the <tt>key</tt> is "a", it will be
+ * treated as 0x0061 rather than as 0x000a</p>
+ * 
+ * <p>Note - there is a significant difference between a unicode codepoint (32 bit int)
+ * and a UTF16 codeunit (=char) - a codepoint consists of one or two codeunits.</p>  
+ * <p>To make sure an int represents a codepoint and not a codeunit, use for example
+ * <code>com.ibm.icu.text.Normalizer2</code> With Nfc mode.
+ * <code>com.ibm.icu.text.UCharacterIterator</code> to retrieve possibly non-BMP codepoints 
+ * from a string.</p>
+ *  
+ * [1] http://java.sun.com/j2se/1.5.0/docs/api/java/util/Properties.html
+ * [2] http://java.sun.com/dtd/properties.dtd
+ *
+ * Class has been migrated from Pipeline 1
+ *
+ *
+ * @author Markus Gylling
+ * @author carl walinder
+ *
+ */
+
+public class UCharReplacer  {
+	private ArrayList<Map<Integer,String>> mSubstitutionTables; 					//ArrayList<HashMap:<int codepoint>,<replaceString>>: all loaded translationtables
+	private Map<Integer,String> mSubstitutionTable = null;						//represents the currently selected translationTable
+	private Iterator<Map<Integer,String>> mSubstitutionTablesIterator = null;			//recycled tables map iterator
+	private Normalizer2 normalizer;
```

Comment:
Not super critical, but I wonder whether we still need the Normalizer2 class. As far as I know you can do NFD normalization with plain Java now. Same for UCharacterIterator. Eliminating dependencies is always a good thing.

Reply:
Hi, i cant find UCharacterIterator in plain java, can you please provide which class you where referring to?

Reply:
Removed dependency to com.ibm.icu